From 735b503bcb09f199473fab0548589fd98b74da07 Mon Sep 17 00:00:00 2001
From: Li Lingfeng <lilingfeng3@huawei.com>
Date: Tue, 15 Nov 2022 16:22:25 +0800
Subject: [PATCH 0050/1398] [Huawei] block: check flags of claimed slave bdev
 to fix uaf for bd_holder_dir

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 187999

--------------------------------

As explained in f4099d89ccff ("block: fix use after free for
bd_holder_dir"), we should make sure the "disk" is still live and
then grab a reference to 'bd_holder_dir'. However, the "disk"
should be "the claimed slave bdev" rather than "the holding disk".

Fixes: f4099d89ccff ("block: fix use after free for bd_holder_dir")
Signed-off-by: Li Lingfeng <lilingfeng3@huawei.com>
---
 fs/block_dev.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/block_dev.c b/fs/block_dev.c
index a9a4eca797d8..886e129f514f 100644
--- a/fs/block_dev.c
+++ b/fs/block_dev.c
@@ -1269,7 +1269,7 @@ int bd_link_disk_holder(struct block_device *bdev, struct gendisk *disk)
 	 * the holder directory.  Hold on to it.
 	 */
 	down_read(&bdev->bd_disk->lookup_sem);
-	if (!(disk->flags & GENHD_FL_UP)) {
+	if (!(bdev->bd_disk->flags & GENHD_FL_UP)) {
 		up_read(&bdev->bd_disk->lookup_sem);
 		return -ENODEV;
 	}
-- 
2.17.1

